File: src/lib/audio/paths.ts
import { type AudioStoragePathInput } from "@/lib/audio/types";

const TRANSLATION_PATTERN = /^[A-Z0-9_-]{2,16}$/u;
const VERSION_PATTERN = /^[a-zA-Z0-9][a-zA-Z0-9._-]{1,127}$/u;
const VOICE_PATTERN = /^[a-z0-9][a-z0-9_-]{1,63}$/u;
const BOOK_PATTERN = /^(?:[1-3][A-Z]{2}|[A-Z]{3})$/u;

export function buildAudioStoragePath({
  translationCode,
  modelVersion,
  voiceId,
  bookId,
  chapter,
  verse,
}: AudioStoragePathInput): string {
  if (!TRANSLATION_PATTERN.test(translationCode)) {
    throw new Error("Invalid audio translation code");
  }
  if (!VERSION_PATTERN.test(modelVersion)) {
    throw new Error("Invalid audio model version");
  }
  if (!VOICE_PATTERN.test(voiceId)) {
    throw new Error("Invalid audio voice ID");
  }
  if (!BOOK_PATTERN.test(bookId)) {
    throw new Error("Invalid audio book ID");
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 999) {
    throw new Error("Invalid audio chapter");
  }
  if (!Number.isInteger(verse) || verse < 1 || verse > 999) {
    throw new Error("Invalid audio verse");
  }

  return [
    translationCode,
    modelVersion,
    voiceId,
    bookId,
    String(chapter).padStart(3, "0"),
    `${String(verse).padStart(3, "0")}.mp3`,
  ].join("/");
}

File: worker/src/onescripture_worker/paths.py
"""Canonical storage paths shared by every generated verse asset."""

from __future__ import annotations

import re

from .errors import PermanentWorkerError
from .models import ClaimedJob

TRANSLATION_PATTERN = re.compile(r"^[A-Z0-9_-]{2,16}$")
VERSION_PATTERN = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9._-]{1,127}$")
VOICE_PATTERN = re.compile(r"^[a-z0-9][a-z0-9_-]{1,63}$")
BOOK_PATTERN = re.compile(r"^(?:[1-3][A-Z]{2}|[A-Z]{3})$")


def build_storage_path(job: ClaimedJob) -> str:
    if not TRANSLATION_PATTERN.fullmatch(job.translation_code):
        raise PermanentWorkerError("invalid_translation", "Invalid translation code")
    if not VERSION_PATTERN.fullmatch(job.model_version):
        raise PermanentWorkerError("invalid_model_version", "Invalid model version")
    if not VOICE_PATTERN.fullmatch(job.voice_id):
        raise PermanentWorkerError("invalid_voice", "Invalid voice identifier")
    if not BOOK_PATTERN.fullmatch(job.book_id):
        raise PermanentWorkerError("invalid_book", "Invalid canonical book identifier")
    if not 1 <= job.chapter <= 999 or not 1 <= job.verse <= 999:
        raise PermanentWorkerError("invalid_reference", "Invalid chapter or verse")
    return (
        f"{job.translation_code}/{job.model_version}/{job.voice_id}/"
        f"{job.book_id}/{job.chapter:03d}/{job.verse:03d}.mp3"
    )
